src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs

```diff
@@ -0,0 +1,280 @@
+                    typeConstraint: LibraryDependencyTarget.Package)
+            };
+
+            await ProjectServices.References.AddOrUpdatePackageReferenceAsync(dependency, token);
+
+            return true;
+        }
+
+        public override Task<bool> UninstallPackageAsync(
+            PackageIdentity packageIdentity, INuGetProjectContext _, CancellationToken token)
+        {
+            ProjectServices.References.RemovePackageReference(packageIdentity.Id);
+
+            return Task.FromResult(true);
+        }
+
+        #endregion
+
+        private string GetBaseIntermediatePath(bool shouldThrow = true)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            var baseIntermediatePath = _vsProjectAdapter.BaseIntermediateOutputPath;
+
+            if (string.IsNullOrEmpty(baseIntermediatePath))
+            {
+                if (shouldThrow)
+                {
+                    throw new InvalidDataException(string.Format(
+                        Strings.BaseIntermediateOutputPathNotFound,
+                        _vsProjectAdapter.FullPath));
+                }
+
+                return null;
+            }
+
+            return baseIntermediatePath;
+        }
+
+        private static PackageReference[] GetPackageReferences(PackageSpec packageSpec)
+        {
+            var frameworkSorter = new NuGetFrameworkSorter();
+
+            return packageSpec
+                .TargetFrameworks
+                .SelectMany(f => GetPackageReferences(f.Dependencies, f.FrameworkName))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+                .ToArray();
+        }
+
+        private static IEnumerable<PackageReference> GetPackageReferences(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework)
+        {
+            return libraries
+                .Where(l => l.LibraryRange.TypeConstraint == LibraryDependencyTarget.Package)
+                .Select(l => ToPackageReference(l, targetFramework));
+        }
+
+        private static PackageReference ToPackageReference(LibraryDependency library, NuGetFramework targetFramework)
+        {
+            var identity = new PackageIdentity(
+                library.LibraryRange.Name,
+                library.LibraryRange.VersionRange.MinVersion);
+
+            return new PackageReference(identity, targetFramework);
+        }
+
+        /// <summary>
+        /// Emulates a JSON deserialization from project.json to PackageSpec in a post-project.json world
+        /// </summary>
+        private async Task<PackageSpec> GetPackageSpecAsync()
+        {
+            await _threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            var projectReferences = await ProjectServices
+                .ReferencesReader
+                .GetProjectReferencesAsync(Common.NullLogger.Instance, CancellationToken.None);
```

Comment:
It'd be nice but this API is a cornerstone of DG restore. And that would greatly enlarge scope of the PR.